File: backend_deno_structured/domain/schedule/event/update.ts
import type { Result } from "../../lang/result.ts";
import type { DateGenerator } from "../../generator.ts";
import type { RepoErr } from "../../repo.ts";
import type { User } from "../user/model.ts";
import type { EventNotFound } from "./read.ts";
import type { EventRepo } from "./repo.ts";
import type { Event } from "./model.ts";
import { ok } from "../../lang/result.ts";
import { eventReadOne } from "./read.ts";

export type EventUpdate = {
    readonly name: Event["name"];
    readonly day: Event["day"];
    readonly begin: Event["begin"];
    readonly end: Event["end"];
    readonly category: Event["category"];
    readonly frequency: Event["frequency"];
    readonly weekendRepeat: Event["weekendRepeat"];
};

export function eventUpdateToEvent(event: EventUpdate, existingEvent: Event, now: Date): Event {
    return {
        id: existingEvent.id,
        name: event.name,
        day: event.day,
        begin: event.begin,
        end: event.end,
        category: event.category,
        frequency: event.frequency,
        weekendRepeat: event.weekendRepeat,
        user: existingEvent.user,
        createdAt: existingEvent.createdAt,
        updatedAt: now,
    };
}

type EventUpdateErrors =
    | RepoErr
    | EventNotFound;

export async function eventUpdate(
    repo: EventRepo,
    dateGenerator: DateGenerator,
    userId: User["id"],
    eventId: Event["id"],
    event: EventUpdate,
): Promise<Result<Event, EventUpdateErrors>> {
    const existingEventResult = await eventReadOne(repo, userId, eventId);
    if (existingEventResult.type === "err") {
        return existingEventResult;
    }
    const now = dateGenerator.gen();
    const builtEvent = eventUpdateToEvent(event, existingEventResult.data, now);
    const updateResult = await repo.cUpdate(builtEvent);
    if (updateResult.type === "err") {
        return updateResult;
    }
    return ok(builtEvent);
}

File: backend/src/domain/schedule/event/update.rs
use crate::domain::generator::DateTimeGen;

use super::{
    error::EventErr,
    model::{Event, EventCategory, EventFrequency},
    repo::EventRepo,
};

#[derive(Debug, PartialEq)]
pub struct EventU {
    pub name: String,
    pub day: String,
    pub begin: String,
    pub end: String,
    pub category: EventCategory,
    pub frequency: Option<EventFrequency>,
    pub weekend_repeat: Option<bool>,
}

pub fn event_from_u(event_u: EventU, id: String, user_id: String, created_at: String) -> Event {
    Event {
        id,
        name: event_u.name,
        day: event_u.day,
        begin: event_u.begin,
        end: event_u.end,
        category: event_u.category,
        frequency: event_u.frequency,
        weekend_repeat: event_u.weekend_repeat,
        user: user_id,
        created_at: created_at.clone(),
        updated_at: created_at,
    }
}

pub fn event_c(
    repo: &dyn EventRepo,
    date_time_gen: &dyn DateTimeGen,
    event_u: EventU,
    event_id: String,
    user_id: String,
) -> Result<Event, EventErr> {
    let old_event = eventReadOne(&repo, &userId, &eventId)?;
    let now = date_time_gen.gen();
    let event = event_from_u(old_event, old_event.data, now);
    repo.cUpdate(event)?;
    return ok(event);
}